### PR TITLE
includ centered fft functions in facade modules and replace FFT functions in other submodules  

### PR DESCRIPTION
**Pull request by junchaoxia**
This PR is addressing the FFT implementation in issue #127 after we developed the facade modules for FFTs. The major improvements are listed as below: 
1) Implement centered FFT functions and unit tests for facade modules.
2) Make all three options (pyfftw, scipy, and cupy) available for all FFT calculations in Image, Volume, Basis, Preprocess, and Cov2D classes. 

To perform corresponding changes in Cov3D submodule, we need to include an additional variable, the length of the transformed axis of the output, into the parameter lists of our customized FFT functions.  This will be updated in another PR.

### CODE REVIEW

File: tests/test_fft.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 import numpy as np
 
 from aspire import config
-from aspire.utils.numeric import fft_class
+from aspire.utils.numeric import fft_object
 
 fft_backends = ['scipy', 'pyfftw']
 numeric_classes = [np, np]
```
Comment (vineetbansal) [inline tests/test_fft.py]: Let's call these `numeric_modules`. Everything else looks great!

Comment (junchaoxia) [reply tests/test_fft.py]: Done.

**Comment by junchaoxia** [comment]
@janden There are ~15 unit tests failed after resolving the conflicts due to the Dtype changes. I will go through them and push back later.

File: src/aspire/basis/ffb_2d.py [hunk 2]
```diff
@@ -230,7 +231,7 @@ def evaluate_t(self, x):
                     self._precomp["gl_weights"][i_r] * self._precomp["gl_nodes"][i_r])
 
         #  1D FFT on the angular dimension for each concentric circle
-        pf = 2 * pi / (2 * n_theta) * fft(pf, 2*n_theta, 2)
+        pf = 2 * pi / (2 * n_theta) * xp.asnumpy(fft.fft(xp.asarray(pf), axis=2))
```
Comment (janden) [inline src/aspire/basis/ffb_2d.py]: Is `axis=2` still needed with the new defaults?

File: src/aspire/image/__init__.py [hunk 3]
```diff
@@ -254,9 +255,10 @@ def _im_translate(self, shifts):
         shifts = shifts.astype(self.dtype)
 
         L = self.res
-        im_f = fft2(im, axes=(1, 2))
-        grid_1d = ifftshift(
-            np.ceil(np.arange(-L/2, L/2, dtype=self.dtype))) * 2 * np.pi / L
+        im_f = xp.asnumpy(fft.fft2(xp.asarray(im), axes=(1, 2)))
```
Comment (janden) [inline src/aspire/image/__init__.py]: Same here, do we need `axes=(1, 2)`?

File: src/aspire/image/__init__.py [hunk 4]
```diff
@@ -266,7 +268,7 @@ def _im_translate(self, shifts):
                         om_y[np.newaxis, :, :] * phase_shifts_y)
         mult_f = np.exp(-1j * phase_shifts)
         im_translated_f = im_f * mult_f
-        im_translated = ifft2(im_translated_f, axes=(1, 2))
+        im_translated = xp.asnumpy(fft.ifft2(xp.asarray(im_translated_f), axes=(1, 2)))
```
Comment (janden) [inline src/aspire/image/__init__.py]: See above.

File: src/aspire/utils/numeric/__init__.py [hunk 5]
```diff
@@ -20,7 +20,7 @@ def fft_class(which):
         from .scipy_fft import ScipyFFT as FFTClass
     else:
         raise RuntimeError(f'Invalid selection for fft class: {which}')
-    return FFTClass
+    return FFTClass()
 
 
-fft = fft_class(config.common.fft)
+fft = fft_object(config.common.fft)
```
Comment (janden) [inline src/aspire/utils/numeric/__init__.py]: Linebreak.

File: src/aspire/utils/numeric/fft.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
+
```
Comment (janden) [inline src/aspire/utils/numeric/fft.py]: This looks like it should lead to a name clash since there's already an `fft` name defined in `__init__`. Somehow it doesn't, but let's not tempt the fates here. A better name would be `base_fft.py` or `_fft.py`.

File: tests/test_fft.py [hunk 7]
```diff
@@ -2,51 +2,75 @@
 import numpy as np
 
 from aspire import config
-from aspire.utils.numeric import fft_class
+from aspire.utils.numeric import fft_object
 
 fft_backends = ['scipy', 'pyfftw']
-numeric_classes = [np, np]
+numeric_modules = [np, np]
```
Comment (janden) [inline tests/test_fft.py:9]: Ideally, this should be done in the same way as for FFT backends (with an object instantiation), but that should go in an issue.

File: tests/test_fft.py [hunk 8]
```diff
@@ -2,51 +2,75 @@
...
             a = xp.random.random(100)
             b = fft.ifftshift(a)
             c = fft.fftshift(b)
             self.assertTrue(xp.allclose(a, c))
+
+    def testCenteredFft(self):
+        for fft, xp in zip(fft_objects, numeric_modules):
+            for nworkers in (-1, 1, 2):
+                a = xp.random.random(100)
+                b = fft.centered_fft(a, workers=nworkers)
```
Comment (janden) [inline tests/test_fft.py:60]: Verify that the center sample of `b` value is equal to the sum of `a`. Same for 2D and 3D.

File: src/aspire/utils/preprocess.py [hunk 9]
```diff
@@ -139,20 +138,26 @@ def downsample(insamples, szout, mask=None):
         # stack of one dimension objects
 
         for idata in range(ndata):
-            insamples_fft = crop_pad(fftshift(fft(insamples[idata])), L_out)*mask
-            outsamples[idata] = np.real(ifft(ifftshift(insamples_fft))*(L_out / L_in))
+            insamples_fft = crop_pad(xp.asnumpy(fft.fftshift(
```
Comment (janden) [inline src/aspire/utils/preprocess.py]: Some of these one-liners become quite hard to parse when we add the `asarray` and `asnumpy` functions. Maybe split up into several lines?

**Review by janden — CHANGES_REQUESTED** [review]

File: src/aspire/basis/ffb_2d.py (hunk 2 above)
Comment (junchaoxia) [reply src/aspire/basis/ffb_2d.py]: Removed in commit 600d977.

File: src/aspire/image/__init__.py (hunk 3 above)
Comment (junchaoxia) [reply src/aspire/image/__init__.py]: Removed in commit 600d977.

File: src/aspire/utils/numeric/__init__.py (hunk 5 above)
Comment (junchaoxia) [reply src/aspire/utils/numeric/__init__.py]: Done in commit 600d977.

File: src/aspire/utils/numeric/fft.py (hunk 6 above)
Comment (junchaoxia) [reply src/aspire/utils/numeric/fft.py]: Renamed to `base_fft.py` in commit ab3cfba.

File: tests/test_fft.py (hunk 8 above)
Comment (junchaoxia) [reply tests/test_fft.py:60]: Done in commit 7763477.

File: src/aspire/utils/preprocess.py (hunk 9 above)
Comment (junchaoxia) [reply src/aspire/utils/preprocess.py]: Changed to multiple lines in commit ff09206.

File: tests/test_fft.py (hunk 7 above)
Comment (junchaoxia) [reply tests/test_fft.py:9]: Created new issue #317.

File: src/aspire/image/__init__.py (hunk 4 above)
Comment (junchaoxia) [reply src/aspire/image/__init__.py]: Done.

File: src/aspire/utils/preprocess.py (hunk 9 above)
Comment (janden) [reply src/aspire/utils/preprocess.py]: They're still single statements, just with many linebreaks in them. What I meant is split it up into several statements and assign to intermediate variables.

Comment (junchaoxia) [reply src/aspire/utils/preprocess.py]: Changed in commit 5164992.

File: src/aspire/utils/preprocess.py [hunk 10]
```diff
@@ -148,23 +148,38 @@ def downsample(insamples, szout, mask=None):
         # stack of one dimension objects
 
         for idata in range(ndata):
-            insamples_fft = crop_pad(fftshift(fft(insamples[idata])), L_out) * mask
-            outsamples[idata] = np.real(ifft(ifftshift(insamples_fft)) * (L_out / L_in))
+            imsample_shifted = fft.fftshift(fft.fft(xp.asarray(insamples[idata])))
```
Comment (janden) [inline src/aspire/utils/preprocess.py]: `insamples_shifted`

File: src/aspire/utils/preprocess.py [hunk 11]
```diff
@@ -148,23 +148,38 @@ def downsample(insamples, szout, mask=None):
...
+            insamples_fft = crop_pad(imsample_shifted, L_out) * mask
+
+            outsammples_shifted = fft.ifft(fft.ifftshift(xp.asarray(insamples_fft)))
```
Comment (janden) [inline src/aspire/utils/preprocess.py]: `outsamples_shifted`

File: src/aspire/image/__init__.py [hunk 12]
```diff
@@ -203,7 +204,13 @@ def downsample(self, ds_res):
         mask = (np.abs(grid["x"]) < ds_res / self.res) & (
             np.abs(grid["y"]) < ds_res / self.res
         )
-        im = np.real(centered_ifft2(centered_fft2(self.data) * mask))
+        im = np.real(
```
Comment (janden) [inline src/aspire/image/__init__.py]: Multiple lines. Same below.

**Review by janden — CHANGES_REQUESTED** [review]

File: src/aspire/utils/preprocess.py (hunk 10 above)
Comment (junchaoxia) [reply src/aspire/utils/preprocess.py]: Done in commit 308ea16.

File: src/aspire/utils/preprocess.py (hunk 11 above)
Comment (junchaoxia) [reply src/aspire/utils/preprocess.py]: Done in commit 308ea16.

File: src/aspire/image/__init__.py (hunk 12 above)
Comment (junchaoxia) [reply src/aspire/image/__init__.py]: Done in commit 308ea16.

File: src/aspire/utils/preprocess.py (hunk 10 above)
Comment (janden) [reply src/aspire/utils/preprocess.py]: Now it says `imsamples_shifted`. It should be `insamples_shifted`.

Comment (junchaoxia) [reply src/aspire/utils/preprocess.py]: Sorry. Changed it and merged it with the previous commit.

**Review by janden — DISMISSED** [review]

**Comment by junchaoxia** [comment]
@janden It took a few hours to solve all conflicts manually and pass the check locally. Let me know if it is OK to merge.

**Review by janden — APPROVED** [review]
